// replicate-api/dist/getModel.js
import { extractModelAndOwner } from "./helpers/extractModelAndOwner.js";
import { makeApiRequest } from "./helpers/makeApiRequest.js";
/** Get information about a model
 * ```typescript
 * const model = await getModel({
 *   model: "stability-ai/stable-diffusion",
 *   token: "Get your token at https://replicate.com/account"
 * })
 *
 * const version = model.version
 * ```
 */
export const getModel = async (options) => {
    const { owner, model } = extractModelAndOwner(options.model);
    const response = await makeApiRequest(options, "GET", `models/${owner}/${model}`);
    const result = {
        url: response.url,
        owner: response.owner,
        name: response.name,
        description: response.description ?? undefined,
        visibility: response.visibility,
        github: response.github_url ?? undefined,
        paper: response.paper_url ?? undefined,
        license: response.license_url ?? undefined,
        version: response.latest_version.id,
    };
    return result;
};
