// replicate-api/dist/getVersions.js
import { extractModelAndOwner } from "./helpers/extractModelAndOwner.js";
import { makeApiRequest } from "./helpers/makeApiRequest.js";
/** Get a list of all versions that are availabe for a model
 * ```typescript
 * const {versions, version} = await getVersions({
 *   model: "stability-ai/stable-diffusion",
 *   token: "Get your token at https://replicate.com/account"
 * })
 * ```
 */
export const getVersions = async (options) => {
    const { owner, model } = extractModelAndOwner(options.model);
    const response = await makeApiRequest(options, "GET", `models/${owner}/${model}/versions`);
    const result = {
        next: response.next,
        previous: response.previous,
        /** The most recent version */
        version: response.results?.[0]?.id,
        versions: response.results?.map(version => ({
            id: version.id,
            created_at: new Date(version.created_at),
            cog_version: version.cog_version,
            schema: version.openapi_schema,
        })),
    };
    return result;
};
